cppguide.md

```diff
@@ -0,0 +1,632 @@
+
+#### Common Variable names
+
+For example:
+
+```cpp
+std::string table_name;  // OK - lowercase with underscore.
+std::string tableName;   // Bad - mixed case.
+```
+
+A generic variable must have the same name as its type. Reducing the number of terms used reduces complexity. It also helps deduce the type of the variable from its name. Most of the time if this rule does not fit well, it's because the name of the type was not well chosen to begin with.
+
+```cpp
+void setTopic(Topic * topic);
+void connect(Database * database);
+
+Point startingPoint, centerPoint;
+Name loginName;
+```
+
+#### Class Data Members
+
+Data members of classes, both static and non-static, are named like ordinary nonmember variables, but with a leading underscore.
+
+```cpp
+class TableInfo {
+  ...
+ private:
+  std::string _table_name;        // OK - leading underscore.
+  static Pool<TableInfo>* _pool;  // OK.
+};
+```
+
+#### Struct Data Members
+
+Data members of structs, both static and non-static, are named like ordinary nonmember variables. They do not have the leading underscores that data members in classes have.
+
+```cpp
+struct UrlTableProperties {
+  std::string name;
+  int num_entries;
+  static Pool<UrlTableProperties>* pool;
+};
+```
+
+See [Structs vs. Classes](#struct-vs-classes) for a discussion of when to use a `struct` versus a `class`.
+
+### Constant Names
+
+Variables declared `constexpr` or `const`, and whose value is fixed for the duration of the program, are named with a leading `k` followed by mixed case. Underscores can be used as separators in the rare cases where capitalization cannot be used for separation. For example:
+
+```cpp
+const int kDaysInAWeek = 7;
+const int kLekaOS1_0_0 = 24;  // LekaOS 1.0.0
+```
+
+All such variables with static storage duration (i.e. statics and globals, see [Storage Duration](https://en.cppreference.com/w/cpp/language/storage_duration#Storage_duration) for details) should be named this way. This convention is optional for variables of other storage classes, e.g. automatic variables, otherwise the usual variable naming rules apply.
+
+### Function Names
+
+Regular functions are written using [camel Case](https://en.wikipedia.org/wiki/Camel_case); accessors and mutators may be named like variables.
+
+Ordinarily, functions should start with a lowercase letter and have a capital letter for each new word.
+
+```cpp
+addTableEntry()
+deleteUrl()
+openFileOrDie()
+```
+
+(The same naming rule applies to class- and namespace-scope constants that are exposed as part of an API and that are intended to look like functions, because the fact that they're objects rather than functions is an unimportant implementation detail.)
+
+Accessors and mutators (get and set functions) may be named like variables if it helps with readability. These often correspond to actual member variables, but this is not required. For example, `int count()`, `int get_count()` and `void set_count(int count)`.
```

Comment:
Is it interesting to add a leading underscore for `protected` and `private` member functions of classes?

It can help to determine which member functions are available outside without checking the access level - `public`, `protected`, `private` - For example:

```cpp
class MyClass {
 public:
  void publicFunction();

 protected:
  void _protectedFunction();

 private
  std::string _name;
};
```

Reply:
If you're using an IDE with auto completion, it will only propose public functions and variables. The protected and private functions and variables won't be accessible or even visible.

If you're not using an IDE with auto completion, you'll have to read the header to know discover the API. Public functions and variables will be at the top. When you encounter private or protected, you know you have to stop reading.

For the library author, the underscore doesn't add anything really besides visual clutter, you should know what you're working on.